File: eel-core/src/test/scala/io/eels/component/avro/AvroSourceTest.scala
package io.eels.component.avro

import java.io.File
import java.nio.file.Paths

import com.typesafe.config.ConfigFactory
import io.eels.{Column, Schema}
import org.apache.avro.util.Utf8
import org.scalatest.{Matchers, OneInstancePerTest, WordSpec}

class AvroSourceTest extends WordSpec with Matchers with OneInstancePerTest {

  import scala.concurrent.ExecutionContext.Implicits.global

  "AvroSource" should {
    "read schema" in {
      val people = AvroSource(Paths.get(new File(getClass.getResource("/test.avro").getFile).getAbsolutePath))
      people.schema shouldBe Schema(List(Column("name"), Column("job"), Column("location")))
    }
    "read strings as java.lang.String when eel.avro.java.string is true" in {
      sys.props.put("eel.avro.java.string", "true")
      ConfigFactory.invalidateCaches()
      val people = AvroSource(Paths.get(new File(getClass.getResource("/test.avro").getFile).getAbsolutePath)).toSet
      people.map(_.values) shouldBe Set(
        List("clint eastwood", "actor", "carmel"),
        List("elton john", "musician", "pinner"),
        List("issac newton", "scientist", "heaven")
      )
    }
    "read strings as utf8 when eel.avro.java.string is false" in {
      sys.props.put("eel.avro.java.string", "false")
      ConfigFactory.invalidateCaches()
      val people = AvroSource(Paths.get(new File(getClass.getResource("/test.avro").getFile).getAbsolutePath)).toSet
      people.map(_.values) shouldBe Set(
        List(new Utf8("clint eastwood"), new Utf8("actor"), new Utf8("carmel")),
        List(new Utf8("elton john"), new Utf8("musician"), new Utf8("pinner")),
        List(new Utf8("issac newton"), new Utf8("scientist"), new Utf8("heaven"))
      )
    }
  }
}

File: eel-core/src/test/kotlin/io/eels/component/avro/AvroSourceTest.kt
package io.eels.component.avro

import java.io.File
import java.nio.file.Paths

import com.typesafe.config.ConfigFactory
import io.eels.schema.Field
import io.eels.schema.Schema
import io.kotlintest.specs.WordSpec
import org.apache.avro.util.Utf8

class AvroSourceTest : WordSpec() {
  init {

    "AvroSource" should {
      "read schema"  {
        val people = AvroSource(Paths.get(File(javaClass.getResource("/test.avro").file).absolutePath))
        people.schema() shouldBe Schema(Field("name", nullable = false), Field("job", nullable = false), Field("location", nullable = false))
      }
      "read strings as java.lang.String when eel.avro.java.string is true"  {
        System.setProperty("eel.avro.java.string", "true")
        ConfigFactory.invalidateCaches()
        val people = AvroSource(Paths.get(File(javaClass.getResource("/test.avro").file).absolutePath)).toFrame(1).toSet()
        people.map { it.values }.toSet() shouldBe setOf(
            listOf("clint eastwood", "actor", "carmel"),
            listOf("elton john", "musician", "pinner"),
            listOf("issac newton", "scientist", "heaven")
        )
        System.setProperty("eel.avro.java.string", "true")
        ConfigFactory.invalidateCaches()
      }
      "read strings as utf8 when eel.avro.java.string is false"  {
        System.setProperty("eel.avro.java.string", "false")
        ConfigFactory.invalidateCaches()
        val people = AvroSource(Paths.get(File(javaClass.getResource("/test.avro").file).absolutePath)).toFrame(1).toSet()
        people.map { it.values }.toSet() shouldBe setOf(
            listOf(Utf8("clint eastwood"), Utf8("actor"), Utf8("carmel")),
            listOf(Utf8("elton john"), Utf8("musician"), Utf8("pinner")),
            listOf(Utf8("issac newton"), Utf8("scientist"), Utf8("heaven"))
        )
        System.setProperty("eel.avro.java.string", "true")
        ConfigFactory.invalidateCaches()
      }
    }
  }
}